x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/TextExpansionQueryBuilder.java

```diff
@@ -198,24 +234,14 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
     }
 
     private QueryBuilder weightedTokensToQuery(String fieldName, TextExpansionResults textExpansionResults) {
-        if (tokenPruningConfig != null) {
-            WeightedTokensQueryBuilder weightedTokensQueryBuilder = new WeightedTokensQueryBuilder(
-                fieldName,
-                textExpansionResults.getWeightedTokens(),
-                tokenPruningConfig
-            );
-            weightedTokensQueryBuilder.queryName(queryName);
-            weightedTokensQueryBuilder.boost(boost);
-            return weightedTokensQueryBuilder;
-        }
-        var boolQuery = QueryBuilders.boolQuery();
-        for (var weightedToken : textExpansionResults.getWeightedTokens()) {
-            boolQuery.should(QueryBuilders.termQuery(fieldName, weightedToken.token()).boost(weightedToken.weight()));
-        }
-        boolQuery.minimumShouldMatch(1);
-        boolQuery.boost(this.boost);
-        boolQuery.queryName(this.queryName);
-        return boolQuery;
+        WeightedTokensQueryBuilder weightedTokensQueryBuilder = new WeightedTokensQueryBuilder(
```

Comment:
This is nice and clean. I think we could think about another optimization here - Right now `WeightedTokensQueryBuilder.toToQuery` pulls the field document count and calculates the token frequency ratio for every query. Since token pruning is an opt in feature for text expansion queries, we might want to update that method in the `WeightedTokenBuilder` to short-circuit this and only get these values if we have a non-null token pruning configuration. WDYT?

Reply:
I think you mean `doToQuery` method here. My understanding is that we should only calculate the token frequency ratio and return BooleanQuery if we have non-null token pruning configuration. Am I right? So, we should go to the following direction:

```
if (this.tokenPruningConfig != null) {
   return new MatchNoDocsQuery("The \"" + getName() + "\" query does not have any pruning configuration");
}

var qb = new BooleanQuery.Builder();
int fieldDocCount = context.getIndexReader().getDocCount(fieldName);
...

```

Please let me know if my understanding is correct about token pruning.

Reply:
You're right, I had a typo - `doToQuery` - here's the [link](https://github.com/elastic/elasticsearch/blob/89c61cde4a1c4003aeacb3740530029f3f5c1504/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/WeightedTokensQueryBuilder.java#L150-L176) to the method I'm talking about. 

Your suggestion to return a `MatchNoDocsQuery` is not what we want here. If we did that, every time we sent in a text expansion query without a pruning configuration, no documents would ever be returned. Since pruning configuration is opt-in and optional this is very undesireable behavior. 

No, what I'm suggesting is altering how we determine whether we want to keep tokens in the `WeightedTokensQueryBuilder`. 

Today, we do the following: 

1. Get the document count for the field name 
2. Calculate the best weight for each token
3. Get the average token frequency ratio
4. Start building a boolean query, determining if we should keep each token. 

If the pruning configuration is null, two things are true: 
1. We want to keep every token, because we don't want any tokens to be pruned
2. Because we want to keep every token there is no need to calculate the counts or frequency ratios. 

So I propose we short-circuit this and only calculate those ratios if there exists a pruning configuration. 

Does this make sense to you?

Reply:
Thank you for the explanation. Now, I got the idea about this optimization. I will change the code and notify you for another review.

Reply:
@kderusso I did some optimization and code clean-up around pruning configuration. Can you please review the changes again? Thank you.